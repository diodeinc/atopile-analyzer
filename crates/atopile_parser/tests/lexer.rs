use atopile_parser::lexer::lex;
use std::fs;

macro_rules! create_lexer_test {
    ($name:ident) => {
        #[test]
        fn $name() {
            let input = fs::read_to_string(concat!(
                "tests/resources/corpus/",
                stringify!($name),
                ".ato"
            ))
            .unwrap();
            let normalized_input = input.replace("\r\n", "\n");
            let output = lex(&normalized_input);
            insta::assert_debug_snapshot!(output);
        }
    };
}

create_lexer_test!(vdivs);
create_lexer_test!(resistors);
create_lexer_test!(transistors);
