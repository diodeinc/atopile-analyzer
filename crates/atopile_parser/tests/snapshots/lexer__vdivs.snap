---
source: crates/atopile_parser/tests/lexer.rs
expression: output
---
(
    [
        Spanned(
            Import,
            0..6,
        ),
        Spanned(
            Name(
                "Resistor",
            ),
            7..15,
        ),
        Spanned(
            From,
            16..20,
        ),
        Spanned(
            String(
                "resistors.ato",
            ),
            21..36,
        ),
        Spanned(
            Newline,
            36..37,
        ),
        Spanned(
            Import,
            37..43,
        ),
        Spanned(
            Name(
                "Capacitor",
            ),
            44..53,
        ),
        Spanned(
            From,
            54..58,
        ),
        Spanned(
            String(
                "capacitors.ato",
            ),
            59..75,
        ),
        Spanned(
            Newline,
            75..76,
        ),
        Spanned(
            Import,
            76..82,
        ),
        Spanned(
            Name(
                "Power",
            ),
            83..88,
        ),
        Spanned(
            From,
            89..93,
        ),
        Spanned(
            String(
                "interfaces.ato",
            ),
            94..110,
        ),
        Spanned(
            Newline,
            110..111,
        ),
        Spanned(
            Import,
            111..117,
        ),
        Spanned(
            Name(
                "Pair",
            ),
            118..122,
        ),
        Spanned(
            From,
            123..127,
        ),
        Spanned(
            String(
                "interfaces.ato",
            ),
            128..144,
        ),
        Spanned(
            Newline,
            144..145,
        ),
        Spanned(
            Newline,
            145..146,
        ),
        Spanned(
            Module,
            146..152,
        ),
        Spanned(
            Name(
                "VDiv",
            ),
            153..157,
        ),
        Spanned(
            From,
            158..162,
        ),
        Spanned(
            Name(
                "_VDiv",
            ),
            163..168,
        ),
        Spanned(
            Colon,
            168..169,
        ),
        Spanned(
            Newline,
            169..170,
        ),
        Spanned(
            Indent,
            170..170,
        ),
        Spanned(
            Name(
                "r_total",
            ),
            174..181,
        ),
        Spanned(
            Colon,
            181..182,
        ),
        Spanned(
            Name(
                "resistance",
            ),
            183..193,
        ),
        Spanned(
            Newline,
            193..194,
        ),
        Spanned(
            Comment(
                " ratio: None",
            ),
            198..211,
        ),
        Spanned(
            Newline,
            211..212,
        ),
        Spanned(
            Name(
                "v_in",
            ),
            216..220,
        ),
        Spanned(
            Colon,
            220..221,
        ),
        Spanned(
            Name(
                "voltage",
            ),
            222..229,
        ),
        Spanned(
            Newline,
            229..230,
        ),
        Spanned(
            Name(
                "v_out",
            ),
            234..239,
        ),
        Spanned(
            Colon,
            239..240,
        ),
        Spanned(
            Name(
                "voltage",
            ),
            241..248,
        ),
        Spanned(
            Newline,
            248..249,
        ),
        Spanned(
            Name(
                "i_q",
            ),
            253..256,
        ),
        Spanned(
            Colon,
            256..257,
        ),
        Spanned(
            Name(
                "current",
            ),
            258..265,
        ),
        Spanned(
            Newline,
            265..266,
        ),
        Spanned(
            Newline,
            266..267,
        ),
        Spanned(
            Name(
                "i_q",
            ),
            271..274,
        ),
        Spanned(
            Equals,
            275..276,
        ),
        Spanned(
            Number(
                "100",
            ),
            277..280,
        ),
        Spanned(
            Name(
                "uA",
            ),
            280..282,
        ),
        Spanned(
            To,
            283..285,
        ),
        Spanned(
            Number(
                "10",
            ),
            286..288,
        ),
        Spanned(
            Name(
                "mA",
            ),
            288..290,
        ),
        Spanned(
            Newline,
            290..291,
        ),
        Spanned(
            Newline,
            291..292,
        ),
        Spanned(
            Comment(
                " assert r_top.value / r_bottom.value within ratio",
            ),
            296..346,
        ),
        Spanned(
            Newline,
            346..347,
        ),
        Spanned(
            Comment(
                " assert r_top.value + r_bottom.value within r_total",
            ),
            351..403,
        ),
        Spanned(
            Newline,
            403..404,
        ),
        Spanned(
            Assert,
            408..414,
        ),
        Spanned(
            Name(
                "v_in",
            ),
            415..419,
        ),
        Spanned(
            Star,
            420..421,
        ),
        Spanned(
            Name(
                "r_bottom",
            ),
            422..430,
        ),
        Spanned(
            Dot,
            430..431,
        ),
        Spanned(
            Name(
                "value",
            ),
            431..436,
        ),
        Spanned(
            Div,
            437..438,
        ),
        Spanned(
            LParen,
            439..440,
        ),
        Spanned(
            Name(
                "r_top",
            ),
            440..445,
        ),
        Spanned(
            Dot,
            445..446,
        ),
        Spanned(
            Name(
                "value",
            ),
            446..451,
        ),
        Spanned(
            Plus,
            452..453,
        ),
        Spanned(
            Name(
                "r_bottom",
            ),
            454..462,
        ),
        Spanned(
            Dot,
            462..463,
        ),
        Spanned(
            Name(
                "value",
            ),
            463..468,
        ),
        Spanned(
            RParen,
            468..469,
        ),
        Spanned(
            Within,
            470..476,
        ),
        Spanned(
            Name(
                "v_out",
            ),
            477..482,
        ),
        Spanned(
            Newline,
            482..483,
        ),
        Spanned(
            Assert,
            487..493,
        ),
        Spanned(
            Name(
                "v_in",
            ),
            494..498,
        ),
        Spanned(
            Div,
            499..500,
        ),
        Spanned(
            LParen,
            501..502,
        ),
        Spanned(
            Name(
                "r_bottom",
            ),
            502..510,
        ),
        Spanned(
            Dot,
            510..511,
        ),
        Spanned(
            Name(
                "value",
            ),
            511..516,
        ),
        Spanned(
            Plus,
            517..518,
        ),
        Spanned(
            Name(
                "r_top",
            ),
            519..524,
        ),
        Spanned(
            Dot,
            524..525,
        ),
        Spanned(
            Name(
                "value",
            ),
            525..530,
        ),
        Spanned(
            RParen,
            530..531,
        ),
        Spanned(
            Within,
            532..538,
        ),
        Spanned(
            Name(
                "i_q",
            ),
            539..542,
        ),
        Spanned(
            Newline,
            542..543,
        ),
        Spanned(
            Newline,
            543..544,
        ),
        Spanned(
            Newline,
            544..545,
        ),
        Spanned(
            Dedent,
            545..545,
        ),
        Spanned(
            Module,
            545..551,
        ),
        Spanned(
            Name(
                "_VDiv",
            ),
            552..557,
        ),
        Spanned(
            Colon,
            557..558,
        ),
        Spanned(
            Newline,
            558..559,
        ),
        Spanned(
            Indent,
            559..559,
        ),
        Spanned(
            Signal,
            563..569,
        ),
        Spanned(
            Name(
                "top",
            ),
            570..573,
        ),
        Spanned(
            Newline,
            573..574,
        ),
        Spanned(
            Signal,
            578..584,
        ),
        Spanned(
            Name(
                "out",
            ),
            585..588,
        ),
        Spanned(
            Newline,
            588..589,
        ),
        Spanned(
            Signal,
            593..599,
        ),
        Spanned(
            Name(
                "bottom",
            ),
            600..606,
        ),
        Spanned(
            Newline,
            606..607,
        ),
        Spanned(
            Newline,
            607..608,
        ),
        Spanned(
            Name(
                "output",
            ),
            612..618,
        ),
        Spanned(
            Equals,
            619..620,
        ),
        Spanned(
            New,
            621..624,
        ),
        Spanned(
            Name(
                "Pair",
            ),
            625..629,
        ),
        Spanned(
            Newline,
            629..630,
        ),
        Spanned(
            Name(
                "output",
            ),
            634..640,
        ),
        Spanned(
            Dot,
            640..641,
        ),
        Spanned(
            Name(
                "io",
            ),
            641..643,
        ),
        Spanned(
            Tilde,
            644..645,
        ),
        Spanned(
            Name(
                "out",
            ),
            646..649,
        ),
        Spanned(
            Newline,
            649..650,
        ),
        Spanned(
            Name(
                "output",
            ),
            654..660,
        ),
        Spanned(
            Dot,
            660..661,
        ),
        Spanned(
            Name(
                "gnd",
            ),
            661..664,
        ),
        Spanned(
            Tilde,
            665..666,
        ),
        Spanned(
            Name(
                "bottom",
            ),
            667..673,
        ),
        Spanned(
            Newline,
            673..674,
        ),
        Spanned(
            Newline,
            674..675,
        ),
        Spanned(
            Name(
                "in",
            ),
            679..681,
        ),
        Spanned(
            Equals,
            682..683,
        ),
        Spanned(
            New,
            684..687,
        ),
        Spanned(
            Name(
                "Power",
            ),
            688..693,
        ),
        Spanned(
            Comment(
                " legacy",
            ),
            694..702,
        ),
        Spanned(
            Newline,
            702..703,
        ),
        Spanned(
            Name(
                "power",
            ),
            707..712,
        ),
        Spanned(
            Equals,
            713..714,
        ),
        Spanned(
            New,
            715..718,
        ),
        Spanned(
            Name(
                "Power",
            ),
            719..724,
        ),
        Spanned(
            Newline,
            724..725,
        ),
        Spanned(
            Name(
                "in",
            ),
            729..731,
        ),
        Spanned(
            Tilde,
            732..733,
        ),
        Spanned(
            Name(
                "power",
            ),
            734..739,
        ),
        Spanned(
            Newline,
            739..740,
        ),
        Spanned(
            Name(
                "power",
            ),
            744..749,
        ),
        Spanned(
            Dot,
            749..750,
        ),
        Spanned(
            Name(
                "vcc",
            ),
            750..753,
        ),
        Spanned(
            Tilde,
            754..755,
        ),
        Spanned(
            Name(
                "top",
            ),
            756..759,
        ),
        Spanned(
            Newline,
            759..760,
        ),
        Spanned(
            Name(
                "power",
            ),
            764..769,
        ),
        Spanned(
            Dot,
            769..770,
        ),
        Spanned(
            Name(
                "gnd",
            ),
            770..773,
        ),
        Spanned(
            Tilde,
            774..775,
        ),
        Spanned(
            Name(
                "bottom",
            ),
            776..782,
        ),
        Spanned(
            Newline,
            782..783,
        ),
        Spanned(
            Newline,
            783..784,
        ),
        Spanned(
            Name(
                "r_top",
            ),
            788..793,
        ),
        Spanned(
            Equals,
            794..795,
        ),
        Spanned(
            New,
            796..799,
        ),
        Spanned(
            Name(
                "Resistor",
            ),
            800..808,
        ),
        Spanned(
            Newline,
            808..809,
        ),
        Spanned(
            Name(
                "r_bottom",
            ),
            813..821,
        ),
        Spanned(
            Equals,
            822..823,
        ),
        Spanned(
            New,
            824..827,
        ),
        Spanned(
            Name(
                "Resistor",
            ),
            828..836,
        ),
        Spanned(
            Newline,
            836..837,
        ),
        Spanned(
            Name(
                "r_top",
            ),
            841..846,
        ),
        Spanned(
            Dot,
            846..847,
        ),
        Spanned(
            Name(
                "package",
            ),
            847..854,
        ),
        Spanned(
            Equals,
            855..856,
        ),
        Spanned(
            String(
                "0402",
            ),
            857..863,
        ),
        Spanned(
            Newline,
            863..864,
        ),
        Spanned(
            Name(
                "r_bottom",
            ),
            868..876,
        ),
        Spanned(
            Dot,
            876..877,
        ),
        Spanned(
            Name(
                "package",
            ),
            877..884,
        ),
        Spanned(
            Equals,
            885..886,
        ),
        Spanned(
            String(
                "0402",
            ),
            887..893,
        ),
        Spanned(
            Newline,
            893..894,
        ),
        Spanned(
            Newline,
            894..895,
        ),
        Spanned(
            Name(
                "top",
            ),
            899..902,
        ),
        Spanned(
            Tilde,
            903..904,
        ),
        Spanned(
            Name(
                "r_top",
            ),
            905..910,
        ),
        Spanned(
            Dot,
            910..911,
        ),
        Spanned(
            Number(
                "1",
            ),
            911..912,
        ),
        Spanned(
            Semicolon,
            912..913,
        ),
        Spanned(
            Name(
                "r_top",
            ),
            914..919,
        ),
        Spanned(
            Dot,
            919..920,
        ),
        Spanned(
            Number(
                "2",
            ),
            920..921,
        ),
        Spanned(
            Tilde,
            922..923,
        ),
        Spanned(
            Name(
                "r_bottom",
            ),
            924..932,
        ),
        Spanned(
            Dot,
            932..933,
        ),
        Spanned(
            Number(
                "1",
            ),
            933..934,
        ),
        Spanned(
            Semicolon,
            934..935,
        ),
        Spanned(
            Name(
                "r_bottom",
            ),
            936..944,
        ),
        Spanned(
            Dot,
            944..945,
        ),
        Spanned(
            Number(
                "2",
            ),
            945..946,
        ),
        Spanned(
            Tilde,
            947..948,
        ),
        Spanned(
            Name(
                "bottom",
            ),
            949..955,
        ),
        Spanned(
            Newline,
            955..956,
        ),
        Spanned(
            Name(
                "r_top",
            ),
            960..965,
        ),
        Spanned(
            Dot,
            965..966,
        ),
        Spanned(
            Number(
                "2",
            ),
            966..967,
        ),
        Spanned(
            Tilde,
            968..969,
        ),
        Spanned(
            Name(
                "out",
            ),
            970..973,
        ),
        Spanned(
            Newline,
            973..974,
        ),
        Spanned(
            Newline,
            974..975,
        ),
        Spanned(
            Dedent,
            975..975,
        ),
        Spanned(
            Module,
            975..981,
        ),
        Spanned(
            Name(
                "VDivLowPassFilter",
            ),
            982..999,
        ),
        Spanned(
            From,
            1000..1004,
        ),
        Spanned(
            Name(
                "VDiv",
            ),
            1005..1009,
        ),
        Spanned(
            Colon,
            1009..1010,
        ),
        Spanned(
            Newline,
            1010..1011,
        ),
        Spanned(
            Indent,
            1011..1011,
        ),
        Spanned(
            Name(
                "cap",
            ),
            1015..1018,
        ),
        Spanned(
            Equals,
            1019..1020,
        ),
        Spanned(
            New,
            1021..1024,
        ),
        Spanned(
            Name(
                "Capacitor",
            ),
            1025..1034,
        ),
        Spanned(
            Newline,
            1034..1035,
        ),
        Spanned(
            Name(
                "cap",
            ),
            1039..1042,
        ),
        Spanned(
            Dot,
            1042..1043,
        ),
        Spanned(
            Name(
                "package",
            ),
            1043..1050,
        ),
        Spanned(
            Equals,
            1051..1052,
        ),
        Spanned(
            String(
                "0402",
            ),
            1053..1059,
        ),
        Spanned(
            Newline,
            1059..1060,
        ),
        Spanned(
            Name(
                "cutoff_frequency",
            ),
            1064..1080,
        ),
        Spanned(
            Colon,
            1080..1081,
        ),
        Spanned(
            Name(
                "frequency",
            ),
            1082..1091,
        ),
        Spanned(
            Newline,
            1091..1092,
        ),
        Spanned(
            Newline,
            1092..1093,
        ),
        Spanned(
            Name(
                "out",
            ),
            1097..1100,
        ),
        Spanned(
            Tilde,
            1101..1102,
        ),
        Spanned(
            Name(
                "cap",
            ),
            1103..1106,
        ),
        Spanned(
            Dot,
            1106..1107,
        ),
        Spanned(
            Number(
                "1",
            ),
            1107..1108,
        ),
        Spanned(
            Semicolon,
            1108..1109,
        ),
        Spanned(
            Name(
                "cap",
            ),
            1110..1113,
        ),
        Spanned(
            Dot,
            1113..1114,
        ),
        Spanned(
            Number(
                "2",
            ),
            1114..1115,
        ),
        Spanned(
            Tilde,
            1116..1117,
        ),
        Spanned(
            Name(
                "bottom",
            ),
            1118..1124,
        ),
        Spanned(
            Newline,
            1124..1125,
        ),
        Spanned(
            Name(
                "cap",
            ),
            1129..1132,
        ),
        Spanned(
            Dot,
            1132..1133,
        ),
        Spanned(
            Name(
                "value",
            ),
            1133..1138,
        ),
        Spanned(
            Equals,
            1139..1140,
        ),
        Spanned(
            Number(
                "100",
            ),
            1141..1144,
        ),
        Spanned(
            Name(
                "nF",
            ),
            1144..1146,
        ),
        Spanned(
            PlusOrMinus,
            1147..1150,
        ),
        Spanned(
            Number(
                "10",
            ),
            1151..1153,
        ),
        Spanned(
            Percent,
            1153..1154,
        ),
        Spanned(
            Newline,
            1154..1155,
        ),
        Spanned(
            Newline,
            1155..1156,
        ),
        Spanned(
            Name(
                "cutoff_frequency",
            ),
            1160..1176,
        ),
        Spanned(
            Colon,
            1176..1177,
        ),
        Spanned(
            Name(
                "frequency",
            ),
            1178..1187,
        ),
        Spanned(
            Newline,
            1187..1188,
        ),
        Spanned(
            Newline,
            1188..1189,
        ),
        Spanned(
            Assert,
            1193..1199,
        ),
        Spanned(
            Number(
                "1",
            ),
            1200..1201,
        ),
        Spanned(
            Div,
            1202..1203,
        ),
        Spanned(
            LParen,
            1204..1205,
        ),
        Spanned(
            Number(
                "2",
            ),
            1205..1206,
        ),
        Spanned(
            Star,
            1207..1208,
        ),
        Spanned(
            Number(
                "3.14",
            ),
            1209..1213,
        ),
        Spanned(
            Star,
            1214..1215,
        ),
        Spanned(
            Name(
                "r_top",
            ),
            1216..1221,
        ),
        Spanned(
            Dot,
            1221..1222,
        ),
        Spanned(
            Name(
                "value",
            ),
            1222..1227,
        ),
        Spanned(
            Star,
            1228..1229,
        ),
        Spanned(
            Name(
                "cap",
            ),
            1230..1233,
        ),
        Spanned(
            Dot,
            1233..1234,
        ),
        Spanned(
            Name(
                "value",
            ),
            1234..1239,
        ),
        Spanned(
            RParen,
            1239..1240,
        ),
        Spanned(
            Within,
            1241..1247,
        ),
        Spanned(
            Name(
                "cutoff_frequency",
            ),
            1248..1264,
        ),
        Spanned(
            Newline,
            1264..1265,
        ),
        Spanned(
            Dedent,
            1265..1265,
        ),
    ],
    [],
)
